thorlcr/activities/indexread/thindexreadslave.cpp

```diff
@@ -558,6 +560,85 @@ class CIndexReadSlaveBase : public CSlaveActivity
+
+            if ((0 == (helper->getFlags() & TIRusesblob)) && !localMerge)
+            {
+                if (!inChildQuery())
+                    useRemoteStreaming = true;
+                else
+                {
+                    /*
+                     * If in a CQ, it is counterproductive to use an index read stream per CQ execution if the index
+                     * involved is relatively small.
+                     * Because, if the index is small and direct reading (and caching) key node pages, it is likely
+                     * that repeated executions will not read any (or few) new key pages (i.e. cache hit).
+                     *
+                     * Example: small 1-way key being remotely read by the whole cluster.
+                     * If it is small it will fit (or mostly fit) in node key cache, and thus mostly read from memory vs over the network etc.
+                     *
+                     */
+
+                    // # data parts excluding TLK if present
+                    unsigned totalNumDataParts = fileDesc.numParts();
+                    if ((totalNumDataParts>1) && !fileDesc.queryProperties().getPropBool("@local"))
+                        totalNumDataParts--; // TLK
+
+                    offset_t logicalFileSize = fileDesc.queryProperties().getPropInt64("@size"); // NB: size is compressed size
+                    if (!logicalFileSize) // not sure when/if this should ever be missing, but..
+                    {
+                        IWARNLOG("Missing @size in meta data for index file '%s'", logicalFilename.get());
+                        // estimate size based on physical size of 1st part
+                        RemoteFilename rfn;
+                        part0.getFilename(0, rfn);
+                        StringBuffer path;
+                        rfn.getPath(path);
+                        Owned<IFile> iFile = createIFile(path);
+                        offset_t partSize = iFile->size();
+                        logicalFileSize = partSize * totalNumDataParts;
+                    }
+
+                    offset_t keyCacheSize = queryJob().getKeyNodeCacheSize() + queryJob().getKeyLeafCacheSize();
+                    offset_t minRemoteCQIndexSizeMb = getOptInt64(THOROPT_MIN_REMOTE_CQ_INDEX_SIZE_MB);
+                    if (minRemoteCQIndexSizeMb)
+                    {
+                        // anything larger is streamed, anything smaller is read directly
+                        if (logicalFileSize > (minRemoteCQIndexSizeMb * 0x100000))
+                            useRemoteStreaming = true;
+                    }
+                    else // no min. size to stream set, so use a heuristic
+                    {
+                        /*
+                         * Rough heuristic.
+                         *
+                         * If (([average compressed part size] * [# parts handling] * [compressionMultiple] * [cacheSizeFitPercentage%]) > [keyCacheSize])
```

Comment:
What is the reason for multiplying by #parts handling rather than #parts?
I would have expected the test to be "does a good proportion of this index fit into the local index cache?"

Reply:
because the cache is per slave, and this slave is only loading key node pages from the parts it has, not the whole key
So the heuristic here is, does a good proportion of the sum [approx.] f the parts I'm handling (often just 1 part), fit in to the cache. In a CQ - it's dealing with all parts, so same thing as whole index.

Reply:
I've added a comment.

Reply:
ok, I was confusing it with a keyed join!